### DISCUSSION TITLE
Using an alias for table names in UPDATE statement

### DISCUSSION THREAD

**Discussion by parvizfarnia** [duckdb/duckdb#7033] [Q&A] [answered]
Hello everyone

I have a question about the [UPDATE](https://duckdb.org/docs/sql/statements/update.html) statement in `DuckDB`. Checking the documentation, if I understand correctly, when you write:

 `UPDATE <table_name> SET <column_name> = ...` 

there is no way to specify an alias after the <table_name> whereas this is possible in `PostgreSQL` (see here the syntax of the [UPDATE](https://www.postgresql.org/docs/15/sql-update.html)). 

Well, this is quite useful in `UPDATE` statements including subqueries where the `WHERE` clause, includes both columns from the table being updated and columns in the subquery and when one or several of these columns **have the same name** so by using an alias one can prevent column name conflict in the query used for `UPDATE`.

I'm going to provide a working test case in `PostgreSQL` so that it might help to show what I am trying to do in `DuckDB`:

```
create table usrs_vars(usrname text, group_id text, varname text);
insert into usrs_vars(usrname, varname) values 
    ('user-01', 'var-01'),
    ('user-01', 'var-02'),
    ('user-01', 'var-03'),
    ('user-01', 'var-04'),
    ('user-01', 'var-05'),
    ('user-02', 'var-03'),
    ('user-02', 'var-04'),
    ('user-03', 'var-01'),
    ('user-04', 'var-01'),
    ('user-04', 'var-04'),
    ('user-05', 'var-03'),
    ('user-05', 'var-04'),
    ('user-05', 'var-05');


# select * from usrs_vars;
 usrname | group_id | varname
---------+----------+---------
 user-01 |          | var-01
 user-01 |          | var-02
 user-01 |          | var-03
 user-01 |          | var-04
 user-01 |          | var-05
 user-02 |          | var-03
 user-02 |          | var-04
 user-03 |          | var-01
 user-04 |          | var-01
 user-04 |          | var-04
 user-05 |          | var-03
 user-05 |          | var-04
 user-05 |          | var-05
(13 lines)
#


create table usrs_groups(usrname text primary key, group_id text);
insert into usrs_groups(usrname, group_id) values 
    ('user-01', 'A'),
    ('user-02', 'A'),
    ('user-03', 'B'),
    ('user-04', 'C'),
    ('user-05', 'A');
    
    
# select * from usrs_groups;
 usrname | group_id
---------+----------
 user-01 | A
 user-02 | A
 user-03 | B
 user-04 | C
 user-05 | A
(5 lines)
#


# As you can see above, the second column 'group_id' in the table 'usrs_vars' 
# is empty (NULL values). I can easily fill for each user, the corresponding 
# value via the following UPDATE statement.

update 
    usrs_vars as t1
set
    group_id = 
    (
        select
            t2.group_id
        from
            usrs_groups as t2
        where
            t1.usrname = t2.usrname
    );
    
    
# select * from usrs_vars;
 usrname | group_id | varname
---------+----------+---------
 user-01 | A        | var-01
 user-01 | A        | var-02
 user-01 | A        | var-03
 user-01 | A        | var-04
 user-01 | A        | var-05
 user-02 | A        | var-03
 user-02 | A        | var-04
 user-03 | B        | var-01
 user-04 | C        | var-01
 user-04 | C        | var-04
 user-05 | A        | var-03
 user-05 | A        | var-04
 user-05 | A        | var-05
(13 lines)
# 
```

Now what I am trying to do is the very same `UPDATE` statement but I don't know as there is no possibility of using alias in `UPDATE` statement in `DuckDB`.

How can I achieve the same ?

Thanks in advance.

**Comment by Mytherin** [comment, marked as answer]
DuckDB does support this in the syntax, perhaps this is only a documentation issue? Does it not work as expected?

**Reply by parvizfarnia** [reply]
Thank you very much. I tried again and indeed this works. I think I probably made a syntax error in my UPDATE statement inside my Python program. I apologize for that and I thank you very much for your help.

As you mentioned, the use of table alias has not been shown explicitly in the documentation. Do you think that I should create a new issue asking for including this in the documentation (as a documentation improvement I mean)?

**Reply by Mytherin** [reply]
Yes, could you please open it in the [documentation repository](https://github.com/duckdb/duckdb-web/issues)?

**Reply by parvizfarnia** [reply]
Done.

[Documenation issue 686](https://github.com/duckdb/duckdb-web/issues/686)

**Reply by madhavanhere** [reply]
I ran in the above issue as well. Turns out the alias for the table being updated **cannot be used in the set section of the statement** 
For example 

Update table1 as a 
set **a,**col1 = b. col1 
from b 
where a.col2 = b.col2

**_is not OK_** 

However 


Update table1 as a 
set **col1** = b. col1 
from b 
where a.col2 = b.col2

is OK. 

Not sure if it was meant to be this way!

**Reply by Mytherin** [reply]
The `SET` columns cannot be qualified, so that is correct. The error here could be more clear, but this also does not work without the `AS`, i.e.:


```sql
D create table tbl(i int, j int);
D UPDATE tbl SET tbl.a=42;
-- Referenced update column tbl not found in table!
```
